File: ai-service/app/serving/common/middleware.py
"""Middleware ASGI thuần: mã yêu cầu, log truy cập, giới hạn thân, bắt lỗi không lường trước.

Gộp vào MỘT middleware để mọi đường (kể cả lỗi 500) đều mang `request_id` và đều ra problem+json — nếu để
Starlette xử lý `Exception` ở lớp ngoài cùng thì contextvar đã bị reset, phản hồi 500 mất request_id.
"""

from __future__ import annotations

import logging
import re
import time
import uuid
from typing import Any

from starlette.datastructures import Headers
from starlette.types import ASGIApp, Message, Receive, Scope, Send

from app.serving.common.logging import request_id_var
from app.serving.common.problem import (
    internal_error,
    problem_response,
    validation_error,
)

HEADER_REQUEST_ID = "x-request-id"
# Chỉ nhận mã do client gửi nếu an toàn để ghi vào log/header (chống chèn dòng log) — giống httpx.RequestID.
_VALID_REQUEST_ID = re.compile(r"^[A-Za-z0-9._-]{8,64}$")
_HEALTH_ROUTES = {"/healthz", "/readyz"}


class ObservabilityMiddleware:
    def __init__(
        self, app: ASGIApp, logger: logging.Logger, max_body_bytes: int
    ) -> None:
        self.app = app
        self.logger = logger
        self.max_body_bytes = max_body_bytes

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        headers = Headers(scope=scope)
        incoming = headers.get(HEADER_REQUEST_ID, "")
        request_id = (
            incoming if _VALID_REQUEST_ID.match(incoming) else str(uuid.uuid4())
        )
        token = request_id_var.set(request_id)
        start = time.perf_counter()
        state: dict[str, Any] = {"status": 500, "bytes": 0, "started": False}

        async def send_wrapper(message: Message) -> None:
            if message["type"] == "http.response.start":
                state["status"] = message["status"]
                state["started"] = True
                raw = list(message.get("headers", []))
                raw.append((b"x-request-id", request_id.encode()))
                raw.append((b"x-content-type-options", b"nosniff"))
                raw.append((b"referrer-policy", b"no-referrer"))
                if not any(k.lower() == b"cache-control" for k, _ in raw):
                    raw.append((b"cache-control", b"no-store"))
                message = {**message, "headers": raw}
            elif message["type"] == "http.response.body":
                state["bytes"] += len(message.get("body", b""))
            await send(message)

        try:
            declared = headers.get("content-length")
            if declared and declared.isdigit() and int(declared) > self.max_body_bytes:
                # Trả 400 (đồng nhất phía Go: thân quá lớn → validation_error), không đọc thân.
                response = problem_response(
                    validation_error("thân request quá lớn"), scope["path"]
                )
                await response(scope, receive, send_wrapper)
            else:
                await self.app(scope, receive, send_wrapper)
        # Chốt chặn cuối: mọi lỗi lạ → 500 chung, chi tiết chỉ vào log.
        except Exception:
            self.logger.exception("unhandled_exception")
            if not state["started"]:
                response = problem_response(internal_error(), scope["path"])
                await response(scope, receive, send_wrapper)
        finally:
            self._log_access(scope, state, start)
            request_id_var.reset(token)

    def _log_access(self, scope: Scope, state: dict[str, Any], start: float) -> None:
        route = scope.get("route")
        route_path = getattr(route, "path", None)
        if not route_path:
            # Tránh bùng nổ nhãn khi bị dò đường dẫn; /healthz, /readyz là route thật nên vẫn có mẫu.
            route_path = "unmatched"
        status = int(state["status"])
        if route_path in _HEALTH_ROUTES:
            level = logging.DEBUG
        elif status >= 500:
            level = logging.ERROR
        elif status >= 400:
            level = logging.WARNING
        else:
            level = logging.INFO
        self.logger.log(
            level,
            "http_request",
            extra={
                "fields": {
                    "method": scope["method"],
                    "route": route_path,
                    "status": status,
                    "duration_ms": int((time.perf_counter() - start) * 1000),
                    "bytes": state["bytes"],
                }
            },
        )

File: backend/pkg/httpx/middleware.go
package httpx

import (
	"fmt"
	"log/slog"
	"net/http"
	"regexp"
	"runtime/debug"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"

	"github.com/NamHaiIT2HUST/DengueSense/backend/pkg/obsx"
)

// HeaderRequestID là tên header truyền mã yêu cầu qua mọi service (docs/09 §6.8).
const HeaderRequestID = "X-Request-ID"

// Chỉ chấp nhận mã do client gửi nếu an toàn để ghi vào log/header (chống chèn dòng log).
var validRequestID = regexp.MustCompile(`^[A-Za-z0-9._-]{8,64}$`)

// RequestID gắn mã yêu cầu (dùng mã đến nếu hợp lệ, không thì sinh UUIDv7) vào context và header phản hồi.
func RequestID() gin.HandlerFunc {
	return func(c *gin.Context) {
		id := c.GetHeader(HeaderRequestID)
		if !validRequestID.MatchString(id) {
			id = uuid.Must(uuid.NewV7()).String()
		}
		c.Header(HeaderRequestID, id)
		c.Request = c.Request.WithContext(obsx.WithRequestID(c.Request.Context(), id))
		c.Next()
	}
}

// SecurityHeaders đặt header bảo mật cơ bản; `Cache-Control: no-store` là mặc định an toàn cho API
// có xác thực — endpoint dữ liệu bất biến (vd GeoJSON theo version) tự ghi đè.
func SecurityHeaders() gin.HandlerFunc {
	return func(c *gin.Context) {
		h := c.Writer.Header()
		h.Set("X-Content-Type-Options", "nosniff")
		h.Set("Referrer-Policy", "no-referrer")
		h.Set("Cache-Control", "no-store")
		c.Next()
	}
}

// MaxBody giới hạn kích thước thân request (docs/09 §6.7).
func MaxBody(limit int64) gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Request.Body = http.MaxBytesReader(c.Writer, c.Request.Body, limit)
		c.Next()
	}
}

// AccessLog ghi 1 dòng log mỗi request. Không ghi query string, header, thân request.
// /healthz và /readyz ở mức debug để không làm ngập log.
func AccessLog(log *slog.Logger) gin.HandlerFunc {
	return func(c *gin.Context) {
		start := time.Now()
		c.Next()

		route := c.FullPath()
		if route == "" {
			route = "unmatched" // tránh bùng nổ nhãn khi bị dò đường dẫn
		}
		status := c.Writer.Status()
		level := slog.LevelInfo
		switch {
		case route == "/healthz" || route == "/readyz":
			level = slog.LevelDebug
		case status >= http.StatusInternalServerError:
			level = slog.LevelError
		case status >= http.StatusBadRequest:
			level = slog.LevelWarn
		}
		log.LogAttrs(c.Request.Context(), level, "http_request",
			slog.String("method", c.Request.Method),
			slog.String("route", route),
			slog.Int("status", status),
			slog.Int64("duration_ms", time.Since(start).Milliseconds()),
			slog.Int("bytes", c.Writer.Size()),
		)
	}
}

// Recover biến panic thành 500 problem+json; chi tiết + stack chỉ ghi vào log, không trả cho client.
func Recover(log *slog.Logger) gin.HandlerFunc {
	return func(c *gin.Context) {
		defer func() {
			if r := recover(); r != nil {
				if r == http.ErrAbortHandler { //nolint:errorlint // sentinel do net/http định nghĩa để dừng handler
					panic(r)
				}
				log.ErrorContext(c.Request.Context(), "panic",
					slog.String("panic", fmt.Sprint(r)),
					slog.String("stack", string(debug.Stack())),
				)
				WriteProblem(c, Internal())
			}
		}()
		c.Next()
	}
}

// NotFoundHandler / MethodNotAllowedHandler trả problem+json thay cho thân mặc định của gin.
func NotFoundHandler() gin.HandlerFunc {
	return func(c *gin.Context) { WriteProblem(c, NotFound("đường dẫn không tồn tại")) }
}

func MethodNotAllowedHandler() gin.HandlerFunc {
	return func(c *gin.Context) { WriteProblem(c, MethodNotAllowed()) }
}

// NewEngine dựng gin.Engine với chuỗi middleware chuẩn của mọi service Go.
// `ContextWithFallback` để *gin.Context truyền như context.Context vẫn đọc được giá trị trong
// request context (request_id, actor) — cần cho strict handler sinh từ oapi-codegen.
func NewEngine(log *slog.Logger, maxBody int64) *gin.Engine {
	gin.SetMode(gin.ReleaseMode)
	r := gin.New()
	r.ContextWithFallback = true
	r.HandleMethodNotAllowed = true
	r.RemoveExtraSlash = false
	r.NoRoute(NotFoundHandler())
	r.NoMethod(MethodNotAllowedHandler())
	r.Use(RequestID(), SecurityHeaders(), AccessLog(log), Recover(log), MaxBody(maxBody))
	return r
}

// TrimBearer tách token khỏi header `Authorization: Bearer <token>`.
func TrimBearer(header string) (string, bool) {
	scheme, token, found := strings.Cut(strings.TrimSpace(header), " ")
	if !found || !strings.EqualFold(scheme, "Bearer") {
		return "", false
	}
	token = strings.TrimSpace(token)
	return token, token != ""
}
